### PR TITLE
3D Overlays can be children

### PR DESCRIPTION
**Pull request by sethalves**
- 3d overlays can be children of entities or avatars

### CODE REVIEW

**Comment by sethalves** [comment]
to test:
- use the Settings menu to enable "Advanced Menus"
- open the console (Edit menu --> Console)
- paste these into the console:
    var boxID = Entities.addEntity({
      type: "Box",
      name: "parent-of-overlay",
      position: Vec3.sum(MyAvatar.position, {x: 0.0, y: 0.4, z: 1.0}),
      dimensions: {x: 0.4, y: 0.4, z: 0.4},
      color: {red: 0, green: 0, blue: 255 },
      dynamic: true
    });

    var highlightBox = Overlays.addOverlay("cube", {
        position: Vec3.sum(MyAvatar.position, {x: 0.0, y: 0.4, z: 1.0}),
        size: 0.8,
        color: { red: 30, green: 255, blue: 30 },
        alpha: 1,
        solid: false,
        visible: true,
        lineWidth: 2.0,
        drawInFront: true,
        parentID: boxID
    });

- close the console.
- bump into the box with your avatar so that it drifts
- observe that the overlay drifts along with the box

**Comment by hifi-gustavo** [comment]
Build console output for commit: 70becfaff1820b08dc73582e7b46bb2920875429 available at: https://gist.github.com/1b8c2df3c587ae150d68c4adfd37cadb<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8339-70becfa.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8339-70becfa.dmg

**Comment by sethalves** [comment]
test this please

**Comment by OmegaHeron** [comment]
Testing Windows 10 - i5 6500 CPU - GTX970 GPU - Desktop and DK2 HMD tested.

Test 1 - Created cube with overlay using script.  Cube is created with overlay, bumping cube sets in motion and overlay green wireframe follows it.

Test 2 - Created overlay attached to AV.  Green wireframe cube appears and is rooted to AV.

For test 1 - deleting cube entity also removes wireframe overlay.

User list, edit windows and other opened overlay dialogs all look same as RC/DEV builds.

On log out/return cube entity from test 1 still present, but, not overlay remains.  I *think* that's correct behavior.

QA Pass unless ^^ is not correct.

**Comment by OmegaHeron** [comment]
One thing that's odd...

After placing a cube using first script.  Every change of domain shows a green wireframe cube at origin (0,0,0) until I restart Interface.

**Comment by sethalves** [comment]
Hm yeah I'm not cleaning up overlays when a parent goes away.  I'll think that over.

**Comment by OmegaHeron** [comment]
Yeah - I see that now.  It seems when I deleted cube and thought overlay deleted, it didn't - it just went to 0,0,0

**Comment by hifi-gustavo** [comment]
Build console output for commit: 70becfaff1820b08dc73582e7b46bb2920875429 available at: https://gist.github.com/1b8c2df3c587ae150d68c4adfd37cadb<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8339-70becfa.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8339-70becfa.dmg

**Comment by sethalves** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Build console output for commit: 28e0ca2e49e7c57fca14cdf7a3259d7a27d8d366 available at: https://gist.github.com/1b8c2df3c587ae150d68c4adfd37cadb<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8339-28e0ca2.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8339-28e0ca2.dmg

**Comment by sethalves** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Build console output for commit: 28e0ca2e49e7c57fca14cdf7a3259d7a27d8d366 available at: https://gist.github.com/1b8c2df3c587ae150d68c4adfd37cadb<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8339-28e0ca2.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8339-28e0ca2.dmg

**Comment by OmegaHeron** [comment]
Ok - for first test, cube entity with "wrapper".  That seems to work correctly - if I delete entity overlayed wireframe is gone - nothing at 0,0,0 etc.  If I don't delete entity and leave domain then return, entity is present but its overlay is not.  If I understand it correctly that's correct in that overlay isn't persisted?

Now a problem.  Second test.  If I trigger it I see attached wireframe, but, if I change domains while in that state I see a wireframe cube at 0,0,0

**Comment by sethalves** [comment]
I see why.  I'll patch again, later.

**Comment by sethalves** [comment]
Okay, now overlays which are children of avatars should stay with the avatar during a domain change (though not through a restart of interface), and children of entities should be deleted during a domain change.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 2c449320d0a46ee23728a0d222b761d5072fa2ba available at: https://gist.github.com/1b8c2df3c587ae150d68c4adfd37cadb<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8339-2c44932.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8339-2c44932.dmg

**Comment by jherico** [comment]
I don't see how to fetch the spatially nestable ID of an overlay.  So how do I parent one overlay to another?

**Comment by sethalves** [comment]
you can't, from this PR.  overlays can be children of entities or avatars, but not of other overlays.  with some more work, overlays could be children of overlays -- the parentFinder thing would have to know how to find overlays.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 1b571f3be3e9cbe148168d82f3aad9905b3ea967 available at: https://gist.github.com/1b8c2df3c587ae150d68c4adfd37cadb<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8339-1b571f3.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8339-1b571f3.dmg

**Comment by OmegaHeron** [comment]
Testing Windows 10

Test 1 - First code snip - creates cube with wireframe overlay.

Cube+overlay creates
Moves with overlay on collide.
Delete cube with edit tools - overlay is deleted (and not left at 0,0,0)
Change domain - return and cube is present, but, overlay is gone.

Test 2 - Second code snip - creates overlay attached to AV.

Overlay persists with AV across domain changes, and as advertised, does not survive interface restart.

This seems to pass now, but, for first test, is it correct that I would lose wireframe after placing cube+overlay then leaving/returning to domain?  If so it passes as other issues found have been corrected.

**Comment by sethalves** [comment]
yes, those are the current expected behaviours.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 607c16849315295403b3987c6d3c71c4a666ee58 available at: https://gist.github.com/1b8c2df3c587ae150d68c4adfd37cadb<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8339-607c168.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8339-607c168.dmg

**Comment by ZappoMan** [comment]
@OmegaHeron - did you do standard FPS performance tests on this PR especially with default scripts running. This PR changes some rendering pipeline issues (using a relative lookup to determine the transform for any 3D overlay) and so I want to make sure it didn't have a negative performance hit.

**Comment by OmegaHeron** [comment]
I'll run the numbers after dinner just to be sure.

**Comment by OmegaHeron** [comment]
Windows 10 - i5 6500 CPU - GTX970 GPU

Desktop mode - 1600x900 maximized window - standard default scripts

**Domain** |localhost
---|---|---|---
**Render Rate**: |145 
**Present Rate**: |1310
**Present New Rate**: |126 
**CPU Buffer Count**: |5413
**GPU Buffer Count**: |1246 
**CPU Memory Usage**: |105771008
**GPU Memory Usage**: |63135744
**CPU Texture Count**: |328
**GPU Texture Count**: |1177
**GPU Transfer Count**: |0
**CPU Texture Memory Usage**: |180771876
**GPU Texture Memory Usage**: |1149566750
**GPU Virtual Memory Usage**: |1289224356
**Triangle Count**: |820434
**Triangle Rate**: |164086800
**API Drawcall Count**: |236
**Drawcall Count**: |239
**Drawcall Rate**: |47800
**State Changes Texture Count**:|266
**Pipeline Count**: |33
**Item Opaques**: |279 |459 |340
**Item Translucents**: |14 |51 |63
**Item Lights**: |21 |80 |35
**Timing Opaques**: |754 |1623 |1042
**Timing Translucents**: |36 |82 |135
**Timing Lighting**: |50 |129 |61
**Render Frame**: |1588 |3422 |2026

Adding First test object -- cube + wireframe

**Domain** |localhost
---|---|---|---
**Render Rate**: |148
**Present Rate**: |1212
**Present New Rate**: |133
**CPU Buffer Count**: |5531
**GPU Buffer Count**: |1681
**CPU Memory Usage**: |107188224
**GPU Memory Usage**: |72798208
**CPU Texture Count**: |328
**GPU Texture Count**: |1307
**GPU Transfer Count**: |0
**CPU Texture Memory Usage**: |51819252
**GPU Texture Memory Usage**: |1321448987
**GPU Virtual Memory Usage**: |1461106593
**Triangle Count**: |828447
**Triangle Rate**: |207111750
**API Drawcall Count**: |181
**Drawcall Count**: |184
**Drawcall Rate**: |46000
**State Changes Texture Count**:|227
**Pipeline Count**: |38
**Item Opaques**: |219
**Item Translucents**: |13
**Item Lights**: |20
**Timing Opaques**: |627
**Timing Translucents**: |36
**Timing Lighting**: |50
**Render Frame**: |1367

Deleted cube + wireframe - added second test - AV attached bit

**Domain** |localhost
---|---|---|---
**Render Rate**: |150
**Present Rate**: |1240
**Present New Rate**: |127
**CPU Buffer Count**: |5535
**GPU Buffer Count**: |1685
**CPU Memory Usage**: |107204608
**GPU Memory Usage**: |72814592
**CPU Texture Count**: |328
**GPU Texture Count**: |1311
**GPU Transfer Count**: |0
**CPU Texture Memory Usage**: |51754548
**GPU Texture Memory Usage**: |1321535083
**GPU Virtual Memory Usage**: |1461192689
**Triangle Count**: |834191
**Triangle Rate**: |119170142
**API Drawcall Count**: |194
**Drawcall Count**: |197
**Drawcall Rate**: |28142
**State Changes Texture Count**:|236
**Pipeline Count**: |38
**Item Opaques**: |233
**Item Translucents**: |13
**Item Lights**: |21
**Timing Opaques**: |628
**Timing Translucents**: |49
**Timing Lighting**: |60
**Render Frame**: |1454

Added back first cube + wireframe and av attached wireframe

**Domain** |localhost
---|---|---|---
**Render Rate**: |149
**Present Rate**: |1184
**Present New Rate**: |123
**CPU Buffer Count**: |5535
**GPU Buffer Count**: |1685
**CPU Memory Usage**: |107204608
**GPU Memory Usage**: |72814592
**CPU Texture Count**: |328
**GPU Texture Count**: |1313
**GPU Transfer Count**: |0
**CPU Texture Memory Usage**: |51705396
**GPU Texture Memory Usage**: |1321600618
**GPU Virtual Memory Usage**: |1461258224
**Triangle Count**: |826599
**Triangle Rate**: |118085571
**API Drawcall Count**: |162
**Drawcall Count**: |166
**Drawcall Rate**: |23714
**State Changes Texture Count**:|211
**Pipeline Count**: |42
**Item Opaques**: |199
**Item Translucents**: |13
**Item Lights**: |20
**Timing Opaques**: |537
**Timing Translucents**: |35
**Timing Lighting**: |51
**Render Frame**: |1296

FPS with items rezzed/attached match closely with numbers recorded for other PR and current RC.  I ran tests at dev-playa and dev-demo but did not attach numbers here - variable/transient AV population made it impossible to present numbers for a 1:1 comparison.

HMD/DK2 shows consistent 75 FPS.

**Comment by imgntn** [comment]
@sethalves this doesn't work for model overlays -- is it possible to extend?  would love that functionality.  otherwise it's a QA pass.  when i parented a box overlay to my model it worked; when i parented a model overlay to my model it did not 
https://gist.github.com/imgntn/f169265828539e5ecd7856cc9a007d80

**Comment by sethalves** [comment]
that's a surprise for me -- I didn't try them.  I'll try to figure out why that is.

**Comment by hifi-gustavo** [comment]
Build console output for commit: c363ca98873e77738969291cee70b3246f37bd42 available at: https://gist.github.com/1b8c2df3c587ae150d68c4adfd37cadb<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8339-c363ca9.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8339-c363ca9.dmg

**Comment by imgntn** [comment]
nice!

File: libraries/shared/src/Transform.h [hunk 1]
```diff
@@ -89,6 +89,10 @@ class Transform {
         return _rotation == other._rotation && _scale == other._scale && _translation == other._translation;
     }
 
+    bool operator!=(const Transform& other) const {
+        return _rotation != other._rotation || _scale != other._scale || _translation != other._translation;
```
Comment (jherico) [inline libraries/shared/src/Transform.h]: wouldn't `!(*this == other)` be just as effective and lower maintenance?

**Pull request merged**